File: 프로그래머스/lv1/131705. 삼총사/삼총사.py
from itertools import combinations

def solution(number) :
    cnt = 0
    for c in combinations(number,3) :
        if sum(c) == 0 :
            cnt += 1
    return cnt

# def dfs(number, data):
#     global cnt

#     if len(data) == 3 and sum(data) == 0:
#         cnt += 1
#         return

#     for i in range(len(number)):
#         dfs(number[i+1:], data+[number[i]])


# def solution(number):
#     global cnt
#     cnt = 0

#     dfs(number, [])

#     return cnt

File: 프로그래머스/1/131705. 삼총사/삼총사.java
import java.util.*;

class Solution {
    
    static int count = 0;
    
    public int solution(int[] number) {
        dfs(number, new ArrayList<>(), 0);
        
        return count;
    }
    
    public void dfs(int[] arr, List<Integer> current, int startIndex) {
        if(current.size() == 3) {
            int sum = current.stream().mapToInt(v -> v).sum();
            if (sum == 0) {
                count += 1;
            }
        }
        
        for (int i=startIndex; i<arr.length; i++) {
            current.add(arr[i]);
            dfs(arr, current, i+1);
            current.remove(current.size() - 1);
        }
    }
}